src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,53 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result=std::monostate>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) { return Result(); }
```

Comment:
Agreed, if your value is not default construct-able it should be trivial to write your particular construction as overrides of `PreVisit`. If all three should be the same you can even

```
template<typename T>
Result PreVisit(const T& _) {
  return Result(special constructor stuff);
}
```

I think this is better than passing around lambdas.

Reply:
I spoke a little loosely in my comment. When I said "hand it a lambda" I really meant "provide it with a function somehow". The actual recommendation that I think makes more sense is to have a

```C++
Result GetDefaultValue() const = 0;
```

in the visitor, which forces the derived classes to indicate how they want to populate default values for the types. It would then be very easy to do this in a derived class for `void`:

```C++
void GetDefaultValue() const override {}
```

Reply:
I think the differences between these are:
- naming (`PreVisit` vs `GetDefaultValue`)
- default behavior (default construction by default vs requiring implementation)
- side effects (i.e. is the 'start value' generated based on a pre visit or a single default value)

Personally, I think having a default constructed value is a good ...default, and this default doesn't preclude start values that are set by Visitor implementations. It just saves typing in the default case.

Additionally, I think it's confusing to have a 'GetDefaultValue' that is different from the `PreVisit` result value.

Interested in @bgogul and @Aferr 's opinions on this though.

Reply:
As I mentioned in a later comment, that isn't the only difference. Certain types that we may wish to return from this visitor (pointers, primitives, types where the default constructor is explicitly deleted) cannot be default constructed. In particular, it is highly idiomatic in C++ to do something akin to returning a sum type by returning a pointer to a base type. Because of that restriction, as currently written, this visitor cannot handle all of the patterns that we may wish to use it for. While ugly, the `GetDefaultValue` method would address this.

Reply:
I should have caught up on the thread before replying. It looks like what Mark is saying is the following :

```C++
class IRTraversingVisitor<...> {
  Result GetDefaultValue() const = 0;
   virtual Result PreVisit(const Module& module) { return GetDefaultResult(); }
 
}
```

Right?  I am even OK with the following:

```C++
class IRTraversingVisitor<...> {
  Result GetDefaultValue() const { return Result() };
   virtual Result PreVisit(const Module& module) { return GetDefaultResult(); }
 
}
```

If the default `GetDefaultValue()` is not sufficient for the `Result` type in question, we will only have to update it in one place?